### PR TITLE
CORE-5530 - Encrypt and decrypt unauthenticated messages

### PR DESCRIPTION
**Pull request by nikinagy**
https://r3-cev.atlassian.net/browse/CORE-5530

Corda API PR: https://github.com/corda/corda-api/pull/649

Salt and aad used for ecies was calculated by following this design: https://github.com/corda/platform-eng-design/blob/bb34527814511663ef8bceb60c17270f45a698ec/core/corda-5/corda-5.0/p2p-ha/e2e-authentication-protocol.md#corda-proposal---ecies-encrypted-message

There was also a crypto bug in `CipherSchemeMetadataProvider`: when we were parsing a public key to PEM string and then converting it back to PublicKey the original algorithm of the key got overridden by the value of 'ECDSA'.

This PR also includes addressing feedback from previous crypto PR: https://github.com/corda/corda-runtime-os/pull/2260#pullrequestreview-1133195125

### CODE REVIEW

**Review by vkolomeyko — DISMISSED** [review]
Looks fine from @corda/http-rpc POV.

I did not review every file, I think my team got looped-in due to changes in E2E area.

File: components/crypto/crypto-service-impl/src/test/kotlin/net/corda/crypto/service/impl/SigningServiceGeneralTests.kt [hunk 1]
```diff
@@ -345,7 +345,7 @@ class SigningServiceGeneralTests {
             scheme = scheme,
             alias = expectedAlias
         )
-        assertSame(generatedKey.publicKey, result)
+        assertEquals(generatedKey.publicKey, result)
```
Comment (yift-r3) [inline components/crypto/crypto-service-impl/src/test/kotlin/net/corda/crypto/service/impl/SigningServiceGeneralTests.kt]: Can we use `assertThat`?

File: components/membership/membership-p2p-impl/src/integrationTest/kotlin/net/corda/membership/impl/p2p/MembershipP2PIntegrationTest.kt [hunk 2]
```diff
@@ -209,6 +258,32 @@ class MembershipP2PIntegrationTest {
         val registrationId = UUID.randomUUID().toString()
         val fakeKey = "fakeKey"
         val fakeSig = "fakeSig"
+
+        val ecdhKey = cryptoOpsClient.generateKeyPair(
+            destination.shortHash.value,
+            "PRE_AUTH",
```
Comment (yift-r3) [inline components/membership/membership-p2p-impl/src/integrationTest/kotlin/net/corda/membership/impl/p2p/MembershipP2PIntegrationTest.kt]: Can we put this in `CryptoConsts.kt`?

File: components/membership/membership-p2p-impl/src/integrationTest/kotlin/net/corda/membership/impl/p2p/MembershipP2PIntegrationTest.kt [hunk 3]
```diff
@@ -209,6 +258,32 @@ class MembershipP2PIntegrationTest {
...
+            destination.shortHash.value + "ecdh",
+            ECDSA_SECP256R1_CODE_NAME
+        )
+
+        val mgm = memberInfoFactory.create(
+            sortedMapOf(
+                GROUP_ID to groupId,
+                PARTY_NAME to destination.x500Name.toString(),
+                Pair(String.format(URL_KEY, "0"), "http://localhost:8080"),
```
Comment (yift-r3) [inline components/membership/membership-p2p-impl/src/integrationTest/kotlin/net/corda/membership/impl/p2p/MembershipP2PIntegrationTest.kt]: https

File: components/membership/membership-p2p-impl/src/integrationTest/kotlin/net/corda/membership/impl/p2p/dummy/TestCryptoOpsClient.kt [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+import org.osgi.service.component.propertytypes.ServiceRanking
+import java.nio.ByteBuffer
+import java.security.KeyPairGenerator
+import java.security.PublicKey
+
+interface TestCryptoOpsClient : CryptoOpsClient
+
+@ServiceRanking(Int.MAX_VALUE)
+@Component(service = [CryptoOpsClient::class, TestCryptoOpsClient::class])
+class TestCryptoOpsClientImpl @Activate constructor(
```
Comment (yift-r3) [inline components/membership/membership-p2p-impl/src/integrationTest/kotlin/net/corda/membership/impl/p2p/dummy/TestCryptoOpsClient.kt]: internal?

File: components/membership/membership-p2p-impl/src/integrationTest/kotlin/net/corda/membership/impl/p2p/dummy/TestEphemeralKeyPairEncryptor.kt [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import net.corda.crypto.ecies.EncryptedDataWithKey
+import net.corda.crypto.ecies.EphemeralKeyPairEncryptor
+import net.corda.v5.crypto.ECDSA_SECP256R1_CODE_NAME
+import org.osgi.service.component.annotations.Activate
+import org.osgi.service.component.annotations.Component
+import org.osgi.service.component.annotations.Reference
+import org.osgi.service.component.propertytypes.ServiceRanking
+import java.security.PublicKey
+
+interface TestEphemeralKeyPairEncryptor : EphemeralKeyPairEncryptor
```
Comment (yift-r3) [inline components/membership/membership-p2p-impl/src/integrationTest/kotlin/net/corda/membership/impl/p2p/dummy/TestEphemeralKeyPairEncryptor.kt:16]: internal?

File: components/membership/membership-p2p-impl/src/integrationTest/kotlin/net/corda/membership/impl/p2p/dummy/TestGroupReaderProvider.kt [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+import net.corda.v5.crypto.PublicKeyHash
+import net.corda.v5.membership.GroupParameters
+import net.corda.v5.membership.MemberInfo
+import net.corda.virtualnode.HoldingIdentity
+import org.osgi.service.component.annotations.Activate
+import org.osgi.service.component.annotations.Component
+import org.osgi.service.component.annotations.Reference
+import org.osgi.service.component.propertytypes.ServiceRanking
+
+interface TestGroupReaderProvider : MembershipGroupReaderProvider {
```
Comment (yift-r3) [inline components/membership/membership-p2p-impl/src/integrationTest/kotlin/net/corda/membership/impl/p2p/dummy/TestGroupReaderProvider.kt:22]: internal?

File: components/membership/membership-p2p-impl/src/main/kotlin/net/corda/membership/impl/p2p/MembershipP2PProcessor.kt [hunk 7]
```diff
@@ -21,10 +24,15 @@ import net.corda.p2p.app.UnauthenticatedMessage
...
 import net.corda.v5.base.exceptions.CordaRuntimeException
 import net.corda.v5.base.util.contextLogger
+import net.corda.v5.cipher.suite.KeyEncodingService
 import java.nio.ByteBuffer
 
 class MembershipP2PProcessor(
-    private val avroSchemaRegistry: AvroSchemaRegistry
+    private val avroSchemaRegistry: AvroSchemaRegistry,
+    private val stableKeyPairDecryptor: StableKeyPairDecryptor,
+    private val keyEncodingService: KeyEncodingService,
```
Comment (yift-r3) [inline components/membership/membership-p2p-impl/src/main/kotlin/net/corda/membership/impl/p2p/MembershipP2PProcessor.kt]: I think we can remove the `private val` from most of them

File: components/membership/membership-p2p-impl/src/main/kotlin/net/corda/membership/impl/p2p/handler/RegistrationRequestHandler.kt [hunk 8]
```diff
@@ -1,41 +1,92 @@
...
+                RegistrationCommand(
+                    StartRegistration(
+                        header.destination,
+                        header.source,
+                        registrationRequest
+                    )
                 )
             )
+        } catch (e: Exception) {
+            logger.warn("Could not create start registration command. Reason: ${e.message}")
```
Comment (yift-r3) [inline components/membership/membership-p2p-impl/src/main/kotlin/net/corda/membership/impl/p2p/handler/RegistrationRequestHandler.kt]: `logger.warn("Could not create start registration command. Reason: ${e.message}", e)`

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationService.kt [hunk 9]
```diff
@@ -163,7 +173,14 @@ class DynamicMemberRegistrationService @Activate constructor(
...
     private val p2pEndpointVerifier = P2pEndpointVerifier(orderVerifier)
 
+    private val headerSerializer: CordaAvroSerializer<UnauthenticatedRegistrationRequestHeader> =
+        cordaAvroSerializationFactory.createAvroSerializer { logger.error("Failed to serialize header.") }
+
+    private val unauthenticatedRegistrationRequestSerializer: CordaAvroSerializer<UnauthenticatedRegistrationRequest> =
+        cordaAvroSerializationFactory.createAvroSerializer { logger.error("Failed to serialize registration request.") }
+
     private var impl: InnerRegistrationService = InactiveImpl
+    private var latestHeader: UnauthenticatedRegistrationRequestHeader? = null
```
Comment (yift-r3) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationService.kt]: We can define this in the function it self, not as a member...

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationService.kt [hunk 10]
```diff
@@ -263,19 +280,42 @@ class DynamicMemberRegistrationService @Activate constructor(
...
+                val data = ephemeralKeyPairEncryptor.encrypt(
+                    mgmKey,
+                    registrationRequestSerializer.serialize(message)!!
+                ) { ek, sk ->
+                    val timestamp = clock.instant().toEpochMilli()
+                    val header = UnauthenticatedRegistrationRequestHeader(1, timestamp, keyEncodingService.encodeAsString(ek))
+                    val serializedHeader = headerSerializer.serialize(header)
+                        ?: throw IllegalArgumentException("Serialized header cannot be null.")
+                    val salt = serializedHeader + keyEncodingService.encodeAsByteArray(sk)
+                    latestHeader = header
```
Comment (yift-r3) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationService.kt]: Can't we change the interface of the `EphemeralKeyPairEncryptor` to return a pair of <EncryptedDataWithKey, T>, so we don't have to use this hack of setting the header from within the block?

File: libs/crypto/cipher-suite-impl/src/test/kotlin/net/corda/cipher/suite/impl/CipherSchemeMetadataTests.kt [hunk 11]
```diff
@@ -531,6 +531,7 @@ class CipherSchemeMetadataTests {
         assert(encodedPublicKey.startsWith("-----BEGIN PUBLIC KEY-----")) { encodedPublicKey }
         assert(encodedPublicKey.contains("-----END PUBLIC KEY-----")) { encodedPublicKey }
         val decodedPublicKey = schemeMetadata.decodePublicKey(encodedPublicKey)
+        assertEquals(keyPair.public.algorithm, decodedPublicKey.algorithm)
```
Comment (yift-r3) [inline libs/crypto/cipher-suite-impl/src/test/kotlin/net/corda/cipher/suite/impl/CipherSchemeMetadataTests.kt:534]: Can we use `assertThat`?

File: libs/crypto/crypto-impl/build.gradle [hunk 12]
```diff
@@ -35,20 +35,6 @@ dependencies {
 tasks.named('jar', Jar) {
     bnd """\
 Import-Package:\
```
Comment (yift-r3) [inline libs/crypto/crypto-impl/build.gradle:37]: IS this part of this change?

File: libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt [hunk 13]
```diff
@@ -175,7 +174,6 @@ class CipherSchemeMetadataProvider : KeyEncodingService {
 
     override fun toSupportedPublicKey(key: PublicKey): PublicKey {
         return when {
-            key::class.java.`package` == BCECPublicKey::class.java.`package` -> key
```
Comment (yift-r3) [inline libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt:178]: I'm not sure I understand this change

**Review by yift-r3 — COMMENTED** [review]
Not sure I understood everything, but it looks OK to me

File: applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt [hunk 14]
```diff
@@ -320,7 +324,7 @@ fun E2eCluster.onboardStaticMembers(groupPolicy: ByteArray) {
         register(
             member.holdingId,
             mapOf(
-                "corda.key.scheme" to KEY_SCHEME
+                "corda.key.scheme" to ECDSA_SECP256R1_CODE_NAME
```
Comment (charlieR3) [inline applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt:327]: Interesting.. We might need to extend this to have separate key scheme for ECDH key. And if we're doing that we should probably allow different key scheme for each key.
Any idea what would happen if a non-EC key is used? i.e. like most people have been using to date.

Comment (charlieR3) [reply applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt:327]: I assume the encryption process requires an EC key.

File: components/membership/registration-impl/build.gradle [hunk 15]
```diff
@@ -9,37 +9,38 @@ description 'Registration service implementations'
 dependencies {
     compileOnly "org.osgi:org.osgi.service.component:$osgiServiceComponentVersion"
     compileOnly 'org.osgi:org.osgi.service.component.annotations'
-    compileOnly "org.osgi:osgi.annotation"
+    compileOnly 'org.osgi:osgi.annotation'
```
Comment (dickon) [inline components/membership/registration-impl/build.gradle]: Best to not change the type of quote. We're still inconsistent, but if we want to clear this up better to keep that in a separate PR.

File: components/membership/membership-p2p-impl/src/main/kotlin/net/corda/membership/impl/p2p/handler/RegistrationRequestHandler.kt [hunk 16]
```diff
@@ -1,41 +1,92 @@
...
+    private fun decryptPayload(payload: ByteBuffer, mgm: HoldingIdentity): ByteArray {
+        val request = avroSchemaRegistry.deserialize<UnauthenticatedRegistrationRequest>(payload)
+        val reqHeader = request.header
+        val memberKey = keyEncodingService.decodePublicKey(reqHeader.key)
+        val mgmKey = getECDHKey(mgm)
+        val serializedReqHeader = headerSerializer.serialize(reqHeader)
+            ?: throw IllegalArgumentException("Serialized header cannot be null.")
+        return stableKeyPairDecryptor.decrypt(
+            mgm.shortHash.value,
+            serializedReqHeader + keyEncodingService.encodeAsByteArray(mgmKey),
```
Comment (charlieR3) [inline components/membership/membership-p2p-impl/src/main/kotlin/net/corda/membership/impl/p2p/handler/RegistrationRequestHandler.kt]: Looking at this line, I'm wondering if there's a reason that the data from `EncryptedDataWithKey` returned from the encrypt function wasn't included in the registration request sent to the MGM?
If the MGM has to calculate the salt again there's a chance it won't match the salt calculated on the member side if the MGM is running a different JVM version or Corda version.
If there's no reason, it might be necessary to update the corda api so that there is an avro version of `EncryptedDataWithKey` and it's included in the body of the registration request.

Comment (charlieR3) [reply components/membership/membership-p2p-impl/src/main/kotlin/net/corda/membership/impl/p2p/handler/RegistrationRequestHandler.kt]: Same for the aad. This is part of `EncryptedDataWithKey` on the member side and calculated separately on the MGM side.

File: components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/SynchronisationIntegrationTest.kt [hunk 17]
```diff
@@ -151,6 +152,9 @@ class SynchronisationIntegrationTest {
         @InjectService(timeout = 5000)
         lateinit var memberInfoFactory: MemberInfoFactory
 
+        @InjectService(timeout = 5000)
+        lateinit var stableKeyPairDecryptor: StableKeyPairDecryptor
```
Comment (charlieR3) [inline components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/SynchronisationIntegrationTest.kt:156]: Just curious - what are the three file changes in this `synchronisation-impl` module for?

File: libs/membership/membership-common/src/main/kotlin/net/corda/membership/lib/MemberInfoExtension.kt [hunk 18]
```diff
@@ -214,5 +215,10 @@ class MemberInfoExtension {
             } else {
                 null
             }
+
+        /** Return the key used for ECIES encryption. */
```
Comment (charlieR3) [inline libs/membership/membership-common/src/main/kotlin/net/corda/membership/lib/MemberInfoExtension.kt]:
```suggestion
        /** Return the key used for ECIES encryption. Only MGMs should have a value set for ecdh key. */
```

File: applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt [hunk 19]
```diff
@@ -21,12 +21,13 @@ import net.corda.membership.httprpc.v1.types.response.RegistrationStatus
 import net.corda.test.util.eventually
 import net.corda.v5.base.util.minutes
 import net.corda.v5.base.util.seconds
+import net.corda.v5.crypto.ECDSA_SECP256R1_CODE_NAME
 import org.assertj.core.api.Assertions.assertThat
 
 const val GATEWAY_CONFIG = "corda.p2p.gateway"
-const val KEY_SCHEME = "CORDA.ECDSA.SECP256R1"
```
Comment (dickon) [inline applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt:27]: I'm not sure why you are specifying the key scheme directly in generateKeyPairIfNotExists rather than having it be a module level constant is a win,  your change improves locality of reference which is probably better anyway.

Just commenting in case I'm missing something subtle with this change.

File: applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt (hunk 14 above)
Comment (nikinagy) [reply applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt:327]: Encryption will fail on the crypto side if you are not using an EC key. Namely a `java.lang.IllegalArgumentException: Keys must use the same algorithm` will be thrown from ECIESUtils.kt.

Comment (charlieR3) [reply applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt:327]: Damn.. We probably need to modify the static registration input then before merging this so the key scheme can differ for keys.

Comment (charlieR3) [reply applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt:327]: Ignore me. Spoke offline. No ecdh key for static so my comment doesn't make sense 🤦‍♂️

File: libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt (hunk 13 above)
Comment (nikinagy) [reply libs/crypto/crypto-impl/src/main/kotlin/net/corda/crypto/impl/CipherSchemeMetadataProvider.kt:178]: This was the 'bug' that was causing algorithm overriding when converting the keys to string and back to public keys. I am not sure either why this fixed it, this was something Alexey suggested.

File: libs/crypto/crypto-impl/build.gradle (hunk 12 above)
Comment (nikinagy) [reply libs/crypto/crypto-impl/build.gradle:37]: Yes, because BCECPublicKey was no longer needed and got removed after the crypto bug fix.

**Review by dickon — DISMISSED** [review]
LGTM, focussing mainly on the crypto pieces since I'm not very familiar with the membership code.

File: applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt (hunk 19 above)
Comment (nikinagy) [reply applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt:27]: I am not sure if i understood correctly this comment. I am basically using a constant (ECDSA_SECP256R1_CODE_NAME) from the crypto APIs which has the same value what KEY_SCHEME had.

File: components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/SynchronisationIntegrationTest.kt (hunk 17 above)
Comment (nikinagy) [reply components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/SynchronisationIntegrationTest.kt:156]: MembershipP2PReadService, which is used in this test, needs StableKeyPairDecryptor to start.

Comment (charlieR3) [reply components/membership/synchronisation-impl/src/integrationTest/kotlin/net/corda/membership/impl/synchronisation/SynchronisationIntegrationTest.kt:156]: Ok, thanks 👍

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationService.kt (hunk 10 above)
Comment (nikinagy) [reply components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationService.kt]: I think it would be better to discuss this, because as far as I remember Alexey wanted that interface as generic and simple as possible. I am not sure we can keep that if we are returning back a header.
Especially since this header is mostly required in our use case as we need to have the same salt and aad values in the MGM side to be able to decrypt the request -> hence we create a header which contains all the info for salt and aad and distribute it to the MGM along with the encrypted request.

File: libs/crypto/cipher-suite-impl/src/test/kotlin/net/corda/cipher/suite/impl/CipherSchemeMetadataTests.kt (hunk 11 above)
Comment (nikinagy) [reply libs/crypto/cipher-suite-impl/src/test/kotlin/net/corda/cipher/suite/impl/CipherSchemeMetadataTests.kt:534]: Normally I would if it were a membership test, but since this is a crypto test and even on the next line they use assertEquals() I think it would look super lame to use here assertThat().

File: applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt [hunk 20]
```diff
@@ -21,12 +21,13 @@ import net.corda.membership.httprpc.v1.types.response.RegistrationStatus
...
 const val P2P_TENANT_ID = "p2p"
 const val HSM_CAT_SESSION = "SESSION_INIT"
+const val HSM_CAT_PRE_AUTH = "PRE_AUTH"
```
Comment (dimosr) [inline applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt:30]: I am probably missing some context here.. What is the `PRE_AUTH` category? Shouldn't that be `ECIES` or something along those lines?

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationService.kt [hunk 21]
```diff
@@ -163,7 +173,14 @@ class DynamicMemberRegistrationService @Activate constructor(
     private val orderVerifier = OrderVerifier()
     private val p2pEndpointVerifier = P2pEndpointVerifier(orderVerifier)
 
+    private val headerSerializer: CordaAvroSerializer<UnauthenticatedRegistrationRequestHeader> =
+        cordaAvroSerializationFactory.createAvroSerializer { logger.error("Failed to serialize header.") }
```
Comment (dimosr) [inline components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationService.kt]: I have noticed the Avro generated classes already have `toByteBuffer`/`fromByteBuffer` methods for serialisation/deserialisation.

Is there an advantage to creating a serialiser, instead?

File: applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt (hunk 20 above)
Comment (nikinagy) [reply applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt:30]: This was something I discussed with only Alexey offline, in his own words:
`it should reflect that it’s being used by P2P only, and I think it’s being used only for “insecure” connections when the member contacting the MGM is not known, the name should reflect that`
So I came up with PRE_AUTH, but I am happy to change it if anyone has any better ideas.

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationService.kt (hunk 21 above)
Comment (charlieR3) [reply components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationService.kt]: I think we need to use the serialization/deserialization factories instead of the `toByteBuffer`/`fromByteBuffer` functions because they use the avro schema registry service which includes some additional metadata in the serialization/deserialization process such as protocol version. 

kdoc from `AvroSchemaRegistry`:
``` kotlin
 * The format of the serialized bytes will be
 * |------+------------+------+------------------|
 *  MAGIC  Fingerprint  Flags  Payload
 *
 *  The fields are defined as:
 *  MAGIC - A constant value which can be used to determine the version of the serialization protocol.
 *  Fingerprint - The unique ID of the schema to which the object was encoded by.
 *  Flags - The set of flags that define characteristics (ie compression) of the payload
 *  Payload - The serialized bytes (possibly compressed) of the object.
```

I remember we ran into issues previously with the byte buffer functions. Looking at old slack messages it was something along the lines of `Error parsing membership message payload
net.corda.v5.base.exceptions.CordaRuntimeException: Could not find data for record with fingerprint: [-26, -80, 46, 79, 61, 65, 108, 105, 99, 101, 44, 32, 76, 61, 76, 111, 110, 100, 111, 110, 44, 32, 67, 61, 71, 66, 12, 65, 66, 67, 49, 50]`
Switching to the factories resolved this and we've been using those factories ever since.

File: applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt (hunk 20 above)
Comment (dimosr) [reply applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt:30]: I see.. so it needs to be more constrained than just the general "ECIES" in the theoretical scenario where the ECIES APIs are used for other use cases in the future?

In that case, `PRE_AUTH` doesn't sound bad. An alternative would be `REGISTRATION` or something like that, but not sure if it's better..

It just struck me, because when talking about pre-auth, we have been referring to pre-auth tokens that we will likely implement later on. So, I was wondering whether it has to do with that or it might collide later on.

File: components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationService.kt (hunk 21 above)
Comment (dimosr) [reply components/membership/registration-impl/src/main/kotlin/net/corda/membership/impl/registration/dynamic/member/DynamicMemberRegistrationService.kt]: ah ok, that makes sense - thanks for explaining.

I think the error you shared might have been due to serialisation and deserialisation not using the same format, e.g. if one of them was using the `toByteBuffer `/ `fromByteBuffer ` approach and the other one the serialisers? If both of them are the same, I believe it should work. I know that, because we have been using the `toByteBuffer `/ `fromByteBuffer ` approach at the p2p in some places.. And your comment made me realise we should probably change these to use the serialisers so that it's self-describing and evolution-friendly.

File: applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt (hunk 20 above)
Comment (nikinagy) [reply applications/workers/release/rpc-worker/src/e2eTest/kotlin/net/corda/applications/workers/rpc/utils/ClusterTestUtils.kt:30]: Yes.

Ah i didn't really know about the pre-auth tokens. We can switch to a better name if you want to not avoid any confusion. I mainly choose PRE_AUTH because the main goal is to register and after the validation you will have authenticated connection with MGM, so this is used before 'authenticated message exchanges'.

**Review by charlieR3 — DISMISSED** [review]

**Review by charlieR3 — APPROVED** [review]

**Review by ronanbrowne — APPROVED** [review]
LGTM (build)

**Review by vkolomeyko — APPROVED** [review]
